clients/drcachesim/tracer/tracer.cpp

```diff
@@ -201,6 +201,87 @@ drmemtrace_custom_module_data(void * (*load_cb)(module_data_t *module),
         return DRMEMTRACE_ERROR;
 }
 
+/* sideline */
+static drvector_t buf_queue;
+static uint queue_size = 0;
+static uint queue_size_mask = 0;
+static volatile uint queue_front = 0;
+static volatile uint queue_tail = 0;
+static void *sideline_exit_event;
+struct buf_entry_t {
+    file_t  file;  // file descriptor, 0 means the end of profiling.
+    void   *data;  // output data pointer, NULL means the end of data for this file.
+    ssize_t size;  // output data size.
+};
+
+static inline void
+trace_buf_enqueue(file_t file, void *data, ssize_t size)
+{
+    bool done = false;
+    buf_entry_t *entry = (buf_entry_t *)dr_global_alloc(sizeof(*entry));
+    entry->file = file;
+    entry->data = data;
+    entry->size = size;
+    do {
+        /* quick racy check without locking */
+        while (((queue_front + 1) & queue_size_mask) == queue_tail) {
+            /* the queue is full, waiting */;
+            dr_thread_yield();
+        }
+        drvector_lock(&buf_queue);
+        if (((queue_front + 1) & queue_size_mask) != queue_tail) {
+            drvector_set_entry(&buf_queue, queue_front, entry);
```

Comment:
So the invariant is that the front is empty?  Maybe a comment on that up above?  How about an assert that the current front file is INVALID_FILE?